To: vim_dev@googlegroups.com
Subject: Patch 7.3.286
Fcc: outbox
From: Bram Moolenaar <Bram@moolenaar.net>
Mime-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
------------

Patch 7.3.286
Problem:    Crash when using "zd" on a large number of folds. (Sam King)
Solution:   Recompute pointer after reallocating array.  Move fewer entries
	    when making room.
Files:	    src/fold.c


*** ../vim-7.3.285/src/fold.c	2010-08-15 21:57:27.000000000 +0200
--- src/fold.c	2011-08-26 16:03:56.000000000 +0200
***************
*** 1469,1479 ****
      }
      else
      {
! 	/* move nested folds one level up, to overwrite the fold that is
  	 * deleted. */
  	moved = fp->fd_nested.ga_len;
  	if (ga_grow(gap, (int)(moved - 1)) == OK)
  	{
  	    /* adjust fd_top and fd_flags for the moved folds */
  	    nfp = (fold_T *)fp->fd_nested.ga_data;
  	    for (i = 0; i < moved; ++i)
--- 1469,1482 ----
      }
      else
      {
! 	/* Move nested folds one level up, to overwrite the fold that is
  	 * deleted. */
  	moved = fp->fd_nested.ga_len;
  	if (ga_grow(gap, (int)(moved - 1)) == OK)
  	{
+ 	    /* Get "fp" again, the array may have been reallocated. */
+ 	    fp = (fold_T *)gap->ga_data + idx;
+ 
  	    /* adjust fd_top and fd_flags for the moved folds */
  	    nfp = (fold_T *)fp->fd_nested.ga_data;
  	    for (i = 0; i < moved; ++i)
***************
*** 1486,1494 ****
  	    }
  
  	    /* move the existing folds down to make room */
! 	    if (idx < gap->ga_len)
  		mch_memmove(fp + moved, fp + 1,
! 					sizeof(fold_T) * (gap->ga_len - idx));
  	    /* move the contained folds one level up */
  	    mch_memmove(fp, nfp, (size_t)(sizeof(fold_T) * moved));
  	    vim_free(nfp);
--- 1489,1497 ----
  	    }
  
  	    /* move the existing folds down to make room */
! 	    if (idx + 1 < gap->ga_len)
  		mch_memmove(fp + moved, fp + 1,
! 				  sizeof(fold_T) * (gap->ga_len - (idx + 1)));
  	    /* move the contained folds one level up */
  	    mch_memmove(fp, nfp, (size_t)(sizeof(fold_T) * moved));
  	    vim_free(nfp);
*** ../vim-7.3.285/src/version.c	2011-08-19 22:28:58.000000000 +0200
--- src/version.c	2011-08-26 16:07:59.000000000 +0200
***************
*** 711,712 ****
--- 711,714 ----
  {   /* Add new patch number below this line */
+ /**/
+     286,
  /**/

-- 
To keep milk from turning sour: Keep it in the cow.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
